pkg/alertdog/alertdog.go

```diff
@@ -0,0 +1,165 @@
+package alertdog
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/PagerDuty/go-pagerduty"
+	"github.com/prometheus/alertmanager/template"
+
+	"github.com/errm/alertdog/pkg/alertmanager"
+)
+
+type Alertmanager interface {
+	Alert(alertmanager.Alert) error
+	Resolve(alertmanager.Alert) error
+}
+
+type Pagerduty interface {
+	ManageEvent(event pagerduty.V2Event) (*pagerduty.V2EventResponse, error)
+}
+
+type Alertdog struct {
+	AlertmanagerEndpoints []string `yaml:"alertmanager_endpoints"`
+	Expected              []*Prometheus
+	CheckInterval         time.Duration
+	Expiry                time.Duration
+	PagerDutyKey          string
+	PagerDutyRunbookURL   string
+
+	mu           sync.RWMutex
+	checkedIn    time.Time
+	alertmanager Alertmanager
+	pagerduty    Pagerduty
+}
+
+func (a *Alertdog) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	defaultInterval, _ := time.ParseDuration("2m")
+	a.CheckInterval = defaultInterval
+	defaultExpiry, _ := time.ParseDuration("5m")
+	a.Expiry = defaultExpiry
+	type plain Alertdog
+	return unmarshal((*plain)(a))
+}
+
+func (a *Alertdog) Setup() {
+	a.alertmanager = alertmanager.Alertmanager{Endpoints: a.AlertmanagerEndpoints, Expiry: a.CheckInterval * 2}
+	a.pagerduty = PagerdutyClient{}
+}
+
+func (a *Alertdog) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	defer r.Body.Close()
+	var data template.Data
+	if err := json.NewDecoder(r.Body).Decode(&data); err != nil {
+		log.Printf("Webhook body invalid, skipping request: %s", err)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	for _, alert := range data.Alerts {
+		a.processWatchdog(alert)
+	}
+	w.WriteHeader(http.StatusOK)
+	return
+}
+
+func (a *Alertdog) processWatchdog(alert template.Alert) {
+	a.CheckIn()
+	for _, prometheus := range a.Expected {
+		var err error
+		switch action := prometheus.CheckIn(alert); action {
+		case ActionAlert:
+			err = a.alertmanager.Alert(prometheus.Alert)
+		case ActionResolve:
+			err = a.alertmanager.Resolve(prometheus.Alert)
+		}
+		if err != nil {
+			a.pagerDutyAlert("alertdog:alertmanager-push", "Alertdog cannot push alerts to alertmanager")
+		}
+	}
+}
+
+func (a *Alertdog) CheckLoop() {
+	checkExpiryTicker := time.NewTicker(a.CheckInterval)
+	for {
+		<-checkExpiryTicker.C
+		a.Check()
```

Comment:
Oh, probably not an issue now, with the rw lock switch.

Reply:
I think this unlikely... since I expect `CheckInterval` to be some number of minutes.

But yeah could be some weird behaviour if `CheckInterval` was very short.

Perhaps it would be more robust to just use time.Sleep here?